src/libs/kata-sys-util/src/mount.rs

```diff
@@ -0,0 +1,1087 @@
+///
+/// For a normal mount, the destination will always be a directory. For bind mount, the destination
+/// must be a directory if the source is a directory, otherwise the destination must be a normal
+/// file. If directories are created, their permissions are initialized to MountPerm.
+///
+/// # Safety
+///
+/// Every container has a root filesystems `rootfs`. When creating bind mounts for a container,
+/// the destination should always be within the container's `rootfs`. Otherwise it's a serious
+/// security flaw for container to read/override host side filesystem contents. Please refer to
+/// following CVEs for example:
+/// - [CVE-2021-30465](https://github.com/opencontainers/runc/security/advisories/GHSA-c3xm-pvg7-gh7r)
+///
+/// To ensure security, the `create_mount_destination()` function takes an extra parameter `root`,
+/// which is used to ensure that `dst` is within the specified directory. And a safe version of
+/// `PathBuf` is returned to avoid TOCTTOU type of flaws.
+pub fn create_mount_destination<S: AsRef<Path>, D: AsRef<Path>, R: AsRef<Path>>(
+    src: S,
+    dst: D,
+    _root: R,
+    fs_type: &str,
+) -> Result<impl AsRef<Path> + Debug> {
+    // TODO: https://github.com/kata-containers/kata-containers/issues/3473
+    let dst = dst.as_ref();
+    let parent = dst
+        .parent()
+        .ok_or_else(|| Error::InvalidPath(dst.to_path_buf()))?;
+    let mut builder = fs::DirBuilder::new();
+    builder.mode(MOUNT_PERM).recursive(true).create(parent)?;
+
+    if fs_type == "bind" {
+        // The source and destination for bind mounting must be the same type: file or directory.
+        if !src.as_ref().is_dir() {
+            fs::OpenOptions::new()
+                .mode(MOUNT_PERM)
+                .write(true)
+                .create(true)
+                .open(dst)?;
+            return Ok(dst.to_path_buf());
+        }
+    }
+
+    if let Err(e) = builder.create(dst) {
+        if e.kind() != std::io::ErrorKind::AlreadyExists {
+            return Err(e.into());
+        }
+    }
+    if !dst.is_dir() {
+        Err(Error::InvalidPath(dst.to_path_buf()))
+    } else {
+        Ok(dst.to_path_buf())
+    }
+}
+
+/// Remount a bind mount into readonly mode.
+///
+/// # Safety
+/// Caller needs to ensure safety of the `dst` to avoid possible file path based attacks.
+pub fn bind_remount_read_only<P: AsRef<Path>>(dst: P) -> Result<()> {
+    let dst = dst.as_ref();
+    if dst.is_empty() {
+        return Err(Error::NullMountPointPath);
+    }
+    let dst = dst
+        .canonicalize()
+        .map_err(|_e| Error::InvalidPath(dst.to_path_buf()))?;
+
+    do_rebind_mount_read_only(dst, MsFlags::empty())
+}
+
+/// Bind mount `src` to `dst` in slave mode, optionally in readonly mode if `readonly` is true.
+///
+/// # Safety
+/// Caller needs to ensure:
+/// - `src` exists.
+/// - `dst` exists, and is suitable as destination for bind mount.
+/// - `dst` is free of file path based attacks.
```

Comment:
imho it's unusual to expect the caller to perform checks - why can't this function do it? Surely that would result in less code overall as the single set of checks would be centralised here?

If not, how about renaming this to `bind_mount_unchecked()` to make it clearer that it doesn't perform checks on the specified params?

Reply:
Renamed